### ISSUE TITLE
Rebroadcast blocks before importing them to reduce the hop latency.

### ISSUE THREAD

**Issue by eskimor** [paritytech/substrate#7926]
@rphmeier suggested this as an optimization. Apparently we have this on Ethereum already, thanks to @tomusdrw.

**Comment by tomusdrw** [comment]
Note that Ethereum implementation is verifying PoW before re-broadcasting, in substrate similarly we should perform external consensus checks (like BABE signature verification) before sending further.

**Comment by rakanalh** [comment]
I talked to @gnunicorn, @tomusdrw and @tomaka about this issue and here's what i gather to start working on this issue:

Currently on Substrate:
- A block announce is received
- Preliminary checks / validations are done
- If valid, the block is downloaded
- Another step if validation on the downloaded block is done
- Transactions are executed in the runtime
- Block is propagated.

The work to be done involves propagating the block **after** the preliminary checks and block download.

**Comment by rphmeier** [comment]
@rakanalh Yes, that sounds accurate. We want all GRANDPA & BABE checks to be done, including those that are done in their respective `ImportBlock` implementations.

**Comment by rakanalh** [comment]
@rphmeier just to confirm, you mean you want all consensus checks to be done before propagating? My initial thought was that we want to only pass the `DefaultBlockAnnounceValidator` validation step, download the block and then propagate the block and only then validate through consensus and execute transaction.

So my plan is more like:
`BlockAnnounce` -> `DefaultBlockAnnounceValidator` -> `Propagate block` -> `Consensus checks` -> `Block runtime / transaction execution`.

**Comment by rphmeier** [comment]
>  just to confirm, you mean you want all consensus checks to be done before propagating?

Yes.

Quoting @tomusdrw from above: 

> in substrate similarly we should perform external consensus checks (like BABE signature verification) before sending further.

The main expensive thing we do with the block is actually executing the block and writing the state to DB. That's what we need to avoid.
